### ISSUE TITLE
properties issues

### ISSUE THREAD

**Issue by farfromrefug** [CartoDB/mobile-carto-libs#16]
I ll sum up a few issues i am seing here:
*text-min-distance: does not seem to do anything . i have tried with very high values (like with places labels) and nothing changes. Is there a special case? What is the unit for it?
* `line-geometry-transform` : 'scale' behave weirdly.i was think that with something like `scale(1.1, 1.1)` i could create some kind of a halo effect but it moves the whole line in totally different places. What is the anchor point for scale? translate seems to be good though.
* could you add support for `line-offset`? this would allow very nice feature rendering like [cliffs](https://wiki.openstreetmap.org/wiki/Tag:natural%3Dcliff) or park ["frontiers"](https://wiki.openstreetmap.org/wiki/Tag:boundary%3Dprotected_area) (like "inside" line of a polygon)
* `line-pattern-file` : kind of render weirdly. The image seems "overly" scaled as shown in the screenshot.
<img width="843" alt="Screenshot 2021-08-17 at 14 52 58" src="https://user-images.githubusercontent.com/655344/129728984-99bcd36c-0930-4308-be18-28d4fc100a3d.png">

**Comment by mtehver** [comment]
Ok, I will try to answer them one by one:

* `text-min-distance`, I will look into this.
* `text-wrap-character`, I think it should be supported, will look into this
* `line-geometry-transform`, scaling works based on local tile coordinates, so the behavior you are describing seems right.
* `line-offset`, I just added support for this. This has been requested before and can be used to create various effects, so this is supported now
* `line-pattern-file` stretching. This has been discussed before, distortion is caused due to keeping line width constant while line stretches when zoomed along the line direction. It would be possible to keep the pattern non-stretched but then zooming causes ugly 'pattern running' artifact which looks very bad to me.

**Comment by farfromrefug** [comment]
@mtehver thanks for your answers i remember now about `line-pattern-file`. I understand what you say. I am just thinking something is wrong as mapbox manage to render them a bit better. By better i mean the "details" are not as "scaled" as in carto. In my screenshot above you can see that we see like a  blurred "circle" when the pattern was actually showing a triangle shape. On Mapbox we still see the shape
<img width="398" alt="Screenshot 2021-08-19 at 09 03 48" src="https://user-images.githubusercontent.com/655344/130023266-eb2b2e81-6ce9-432c-8ba6-9cb30180a30d.png">

**Comment by farfromrefug** [comment]
@mtehver just tested your latest changes, pretty awesome!!! still a weird behavior with `text-wrap-width`
as you can see in the video it seems to not always wrap at the same "place".
I use this:
```
text-wrap-character: '-';
text-wrap-width:step([view::zoom], (9, 100), (15, 180));
text-wrap-before:true;
```

https://user-images.githubusercontent.com/655344/130239718-5e082e5d-297e-40f5-9c39-b6d334491092.mp4

**Comment by farfromrefug** [comment]
Time to show the progress. It is going really well in copying Mapbox style.
![Screenshot_1629473300](https://user-images.githubusercontent.com/655344/130257040-7a31fd25-da1c-42f7-8203-ba0257ee9d03.png)
![Screenshot_1629473327](https://user-images.githubusercontent.com/655344/130257061-c7a9d1dc-44e6-4734-a41a-503af009b22f.png)

**Comment by farfromrefug** [comment]
And this is my osm variant. Just a variant of the same style.
![Screenshot_1629473454](https://user-images.githubusercontent.com/655344/130257277-352061a8-1b48-4c52-bd20-fc3a65585ef8.png)

![Screenshot_1629473438](https://user-images.githubusercontent.com/655344/130257261-0a30c577-c0c5-471a-a38a-d4a663654041.png)

**Comment by mtehver** [comment]
Your styles look great! Any chance of making them publicly available in github?

**Comment by farfromrefug** [comment]
@mtehver yes i was actually hoping you would ask for this :D Let me clean it up a bit and i ll create a repo just for it.

**Comment by farfromrefug** [comment]
@mtehver here you go https://github.com/farfromrefug/new_carto_theme

**Comment by farfromrefug** [comment]
@mtehver it seems there is an issue with `line-offset`. I dont see it with 'cliffs' when i use it but with this code it draws the "offset" line twice, one on each side
<img width="707" alt="Screenshot 2021-08-21 at 21 07 24" src="https://user-images.githubusercontent.com/655344/130332479-ce0b3820-dcc2-4f36-9a67-d9f8fbcb108d.png">

Another side effect of `line-offset`, on certain zooms  i see unwanted lines (same code)
<img width="844" alt="Screenshot 2021-08-21 at 21 11 41" src="https://user-images.githubusercontent.com/655344/130332568-3085d0eb-b66e-4ef0-951d-25c977c8ae8f.png">

**Comment by mtehver** [comment]
@farfromrefug What type of geometry do you use with offsetted lines? In case of polygons, this could happen at tile borders if you offset lines 'inwards'. The reason is that polygons are clipped at tile boundaries (with a small extra buffer usually) and when you shift the lines inwards, this clipping becomes visible. Thus this is an artifact of tiling and can not be really fixed (other than increasing the buffer size).

**Comment by farfromrefug** [comment]
@mtehver it makes sense and yes those are polygons (parks). I ll see what i can do on my side
